tensorflow_addons/layers/crf.py

```diff
@@ -0,0 +1,478 @@
+            tf.keras.regularizers.serialize(self.bias_regularizer),
+            "kernel_constraint":
+            tf.keras.constraints.serialize(self.kernel_constraint),
+            "chain_constraint":
+            tf.keras.constraints.serialize(self.chain_constraint),
+            "boundary_constraint":
+            tf.keras.constraints.serialize(self.boundary_constraint),
+            "bias_constraint":
+            tf.keras.constraints.serialize(self.bias_constraint)
+        }
+        base_config = super(CRF, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def compute_output_shape(self, input_shape):
+        output_shape = input_shape[:2]
+        return output_shape
+
+    def compute_mask(self, input_, mask=None):
+        """keep mask shape [batch_size, max_seq_len]"""
+        return mask
+
+    def get_negative_log_likelihood(self, y_true):
+        # TODO(howl-anderson): remove unnecessary typing cast
+        self.potentials = tf.keras.backend.cast(self.potentials, tf.float32)
+        y_true = tf.keras.backend.cast(y_true, tf.int32)
+        self.sequence_length = tf.keras.backend.cast(self.sequence_length,
+                                                     tf.int32)
+        # self.chain_kernel = tf.keras.backend.cast(self.chain_kernel,
+        #                                           tf.float32)
+
+        log_likelihood, _ = crf_log_likelihood(
+            self.potentials, y_true, self.sequence_length, self.chain_kernel)
+
+        return -log_likelihood
+
+    def get_loss(self, y_true, y_pred):
+        # we don't use y_pred, but caller pass it anyway, ignore it
+        return self.get_negative_log_likelihood(y_true)
+
+    def get_accuracy(self, y_true, y_pred):
+        judge = tf.keras.backend.cast(
+            tf.keras.backend.equal(y_pred, y_true), tf.keras.backend.floatx())
+        if self.mask is None:
+            return tf.keras.backend.mean(judge)
+        else:
+            mask = tf.keras.backend.cast(self.mask, tf.keras.backend.floatx())
+            return (tf.keras.backend.sum(judge * mask) /
+                    tf.keras.backend.sum(mask))
+
+    def _dense_layer(self, input_):
+        if self.use_kernel:
+            output = self.activation(
+                tf.keras.backend.dot(input_, self.kernel) + self.bias)
+        else:
+            output = input_
+
+        return tf.keras.backend.cast(output, self.chain_kernel.dtype)
+
+    def __call__(self, inputs, *args, **kwargs):
+        outputs = super(CRF, self).__call__(inputs, *args, **kwargs)
+
+        # A hack that add _keras_history to EagerTensor, make it more like normal Tensor
+        for tensor in tf.nest.flatten(outputs):
+            if not hasattr(tensor, '_keras_history'):
+                tensor._keras_history = (self, 0, 0)
+
+        return outputs
+
+    @property
+    def _compute_dtype(self):
+        # fixed output dtype from underline CRF functions
+        return tf.int32
```

Comment:
I don't see this used anywhere?

Reply:
Unless it's overriding `tf.keras.layers.Layer`

Reply:
Yes, it overriding the base layer's method. CRF layer's output dtype is int32, but default dtype of layer is tf.keras.backend.floatx()